src/feature/rend/rendclient.c

```diff
@@ -1253,47 +1253,61 @@ rend_parse_service_authorization(const or_options_t *options,
 /** The given circuit is being freed. Take appropriate action if it is of
  * interest to the client subsystem. */
 void
-rend_client_circuit_cleanup(circuit_t *circ)
+rend_client_circuit_cleanup(const circuit_t *circ)
 {
-  int reason = circ->marked_for_close_reason;
-  int orig_reason = circ->marked_for_close_orig_reason;
+  int reason, orig_reason;
+  bool has_timed_out, ip_is_redundant;
+  const origin_circuit_t *ocirc = NULL;
 
   tor_assert(circ);
+  tor_assert(CIRCUIT_IS_ORIGIN(circ));
 
-  if (circ->purpose == CIRCUIT_PURPOSE_C_INTRODUCE_ACK_WAIT) {
-    origin_circuit_t *ocirc = TO_ORIGIN_CIRCUIT(circ);
-    int timed_out = (reason == END_CIRC_REASON_TIMEOUT);
+  reason = circ->marked_for_close_reason;
+  orig_reason = circ->marked_for_close_orig_reason;
+  ocirc = CONST_TO_ORIGIN_CIRCUIT(circ);
+
+  has_timed_out = (reason == END_CIRC_REASON_TIMEOUT);
+  ip_is_redundant = (orig_reason == END_CIRC_REASON_IP_NOW_REDUNDANT);
+
+  switch (circ->purpose) {
+  case CIRCUIT_PURPOSE_C_INTRODUCE_ACK_WAIT:
```

Comment:
Fully on board with this refactoring as well.

However, one behavior-change I can see is that the old code used to check for `ocirc->rend_data` here whereas the new code takes it for granted in both cases. Let's make sure that this does not cause any crashes.

Reply:
The idea behind this is that if you reach the HSv2 subsystem, you are _certain_ that it is a v2 circuit. In other words, HSv2 entry point should be reduced in favor of the `hs_circ_cleanup()` entry point. And thus I removed the check.

I could bring it back with `BUG()` ?

Reply:
Sounds good.